### DISCUSSION TITLE
[Question] How can the performance speed be increased on the EV3?

### DISCUSSION THREAD

**Discussion by PhilKu7** [pybricks/support#448] [Q&A] [answered]
**Question**
Is there a way to speed up loops in pybricks-micropython on the EV3 and making them more steady?

**Context**
I'm participating in WRO and need the robot to stop at the line, when it detects it. This means I need a loop to execute about every 5ms. If a loop takes longer than that, it would mean that the robot might not see the line, because it couldn't make a new sample. 

**Screenshots**
![image](https://user-images.githubusercontent.com/86799686/130805132-c1527271-478a-4b25-9967-de046f42d607.png)
This is the output of the following program to show the time taken per loop.

```Python
#!/usr/bin/env pybricks-micropython
from pybricks.tools import wait, StopWatch, DataLog

all_times = []
data = DataLog("time", "loop time", name="loop_measurements", timestamp=False)

watch = StopWatch()
watch.reset()

# measure the time for 1000 loops
for n in range(1000):
    all_times.append(watch.time())

last_time = 0
for n in all_times:
    data.log(n, n-last_time)
    last_time = n
```
In this experiment the EV3 brick had 3 medium motors, 3 color sensor and an ultrasonic sensor connected. It had Bluetooth disabled and was not connected by USB, when the program was running. The battery had a voltage of 7.14mV.
I'm using the EV3 with the ev3dev image from Lego Education (v2.0) and pybricks-micropython v3.0 (updated as described in issue #370).

Is there any workaround to increase the speed to have all loops faster than 5ms?

**Comment by dlech** [comment, marked as answer]
We were actually just discussing this... 6 years ago. Seems like yesterday. :laughing: See https://github.com/ev3dev/ev3dev/issues/324. The graph looks very familiar.

The short answer is that Linux is not a realtime operating system, so this is just how ev3dev is. So you will have to look for solutions that work around this. For example, maybe it is possible to use more than one sensor or a larger sensor (Mindsensors light array?) to be able to detect the line while the robot is still moving fast. Or maybe you can make the robot go slower when you think you are getting close to the line?

**Reply by PhilKu7** [reply]
Thank you for your fast reply.
I saw this discussion before, but thought that there would be a new way to solve it. I also use ROBOTC and as far as I know it also works based on Linux, but you can reach much faster loops (so my test loops all took under 1ms). I imagine that this cannot be applied to ev3dev so I will have to try and slow down the robot before lines.
